src/backend/renderer/mod.rs

```diff
@@ -0,0 +1,287 @@
+}
+
+impl Transform {
+    /// A projection matrix to apply this transformation
+    pub fn matrix(&self) -> Matrix3<f32> {
+        match self {
+            Transform::Normal => Matrix3::new(1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0),
+            Transform::_90 => Matrix3::new(0.0, -1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0),
+            Transform::_180 => Matrix3::new(-1.0, 0.0, 0.0, 0.0, -1.0, 0.0, 0.0, 0.0, 1.0),
+            Transform::_270 => Matrix3::new(0.0, 1.0, 0.0, -1.0, 0.0, 0.0, 0.0, 0.0, 1.0),
+            Transform::Flipped => Matrix3::new(-1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0),
+            Transform::Flipped90 => Matrix3::new(0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0),
+            Transform::Flipped180 => Matrix3::new(1.0, 0.0, 0.0, 0.0, -1.0, 0.0, 0.0, 0.0, 1.0),
+            Transform::Flipped270 => Matrix3::new(0.0, -1.0, 0.0, -1.0, 0.0, 0.0, 0.0, 0.0, 1.0),
+        }
+    }
+
+    /// Inverts any 90-degree transformation into 270-degree transformations and vise versa.
+    ///
+    /// Flipping is preserved and 180/Normal transformation are uneffected.
+    pub fn invert(&self) -> Transform {
+        match self {
+            Transform::Normal => Transform::Normal,
+            Transform::Flipped => Transform::Flipped,
+            Transform::_90 => Transform::_270,
+            Transform::_180 => Transform::_180,
+            Transform::_270 => Transform::_90,
+            Transform::Flipped90 => Transform::Flipped270,
+            Transform::Flipped180 => Transform::Flipped180,
+            Transform::Flipped270 => Transform::Flipped90,
+        }
+    }
+
+    /// Transformed size after applying this transformation.
+    pub fn transform_size(&self, width: u32, height: u32) -> (u32, u32) {
+        if *self == Transform::_90
+            || *self == Transform::_270
+            || *self == Transform::Flipped90
+            || *self == Transform::Flipped270
+        {
+            (height, width)
+        } else {
+            (width, height)
+        }
+    }
+}
+
+#[cfg(feature = "wayland-frontend")]
+impl From<wayland_server::protocol::wl_output::Transform> for Transform {
+    fn from(transform: wayland_server::protocol::wl_output::Transform) -> Transform {
+        use wayland_server::protocol::wl_output::Transform::*;
+        match transform {
+            Normal => Transform::Normal,
+            _90 => Transform::_90,
+            _180 => Transform::_180,
+            _270 => Transform::_270,
+            Flipped => Transform::Flipped,
+            Flipped90 => Transform::Flipped90,
+            Flipped180 => Transform::Flipped180,
+            Flipped270 => Transform::Flipped270,
+        }
+    }
+}
+
+/// Abstraction for Renderers, that can render into different targets
+pub trait Bind<Target>: Unbind {
+    /// Bind a given rendering target, which will contain the rendering results until `unbind` is called.
+    fn bind(&mut self, target: Target) -> Result<(), <Self as Renderer>::Error>;
```

Comment:
This should document what behavior is expected if `bind()` is called twice with no `unbind()` in between. Is this an error, or will the latest target just replace the previous one?

Reply:
I have no idea, how this will work with future renderers, but for gles2 the current target is just replaced.

So the question is: Do we wanna be cautious here and define that as an `Error` or renderer-dependent or do we roll with gles2 behavior and potentially fix that up later? (e.g. when we introduce a vulkan renderer.)

Reply:
If we don't want to commit to any behavior, we can document that doing it is generally unadvised and that the actual behavior is renderer-dependent, but it should be documented clearly I think.